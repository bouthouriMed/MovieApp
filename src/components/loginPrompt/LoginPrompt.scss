@import "@/styles/variables.scss";
@import "@/styles/mixins.scss";

.login-prompt {
  min-height: 60vh;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 2rem;

  .card {
    background: rgba(18, 18, 18, 0.9);
    border-radius: 16px;
    padding: 3rem 2rem;
    text-align: center;
    color: $text-color;
    max-width: 450px;
    width: 100%;
    box-shadow: 0 10px 30px rgba(0, 0, 0, 0.5);
    backdrop-filter: blur(6px);
    transition: transform 0.3s ease, box-shadow 0.3s ease;

    &:hover {
      transform: translateY(-5px);
      box-shadow: 0 15px 35px rgba(0, 0, 0, 0.6);
    }

    h2 {
      font-family: $font-heading;
      font-size: 2rem;
      margin-bottom: 1rem;
      color: $primary-color;
    }

    p {
      font-family: $font-body;
      font-size: 1rem;
      margin-bottom: 2rem;
      line-height: 1.5;
      color: lighten($text-color, 30%);
    }

    .login-btn {
      background: linear-gradient(135deg, #6c63ff, #ff6584);
      color: $text-color;
      border: none;
      border-radius: 30px;
      padding: 0.8rem 2rem;
      font-weight: 600;
      font-size: 1rem;
      cursor: pointer;
      box-shadow: 0 6px 15px rgba(108, 99, 255, 0.4);
      font-family: $font-body;
      letter-spacing: 0.5px;
      transition: all 0.3s ease;

      &:hover {
        filter: brightness(1.15);
        transform: translateY(-2px);
      }

      &:active {
        transform: translateY(0);
        filter: brightness(1);
      }
    }
  }
}
