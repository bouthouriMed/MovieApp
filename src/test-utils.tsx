import React, { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore, Reducer } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import authReducer, { AuthState } from "./store/authSlice";
import watchListReducer, { WatchListState } from "./store/watchListSlice";
import { moviesApi } from "./store/movieSlice";

// --- Helper for renderHook with optional store ---
import {
  renderHook as rtlRenderHook,
  RenderHookOptions,
} from "@testing-library/react";
import { userApi } from "./store/userSlice";

// Explicit reducers type
export const reducers: {
  auth: Reducer<AuthState>;
  watchList: Reducer<WatchListState>;
  [moviesApi.reducerPath]: typeof moviesApi.reducer;
  [userApi.reducerPath]: typeof userApi.reducer;
} = {
  auth: authReducer,
  watchList: watchListReducer,
  [moviesApi.reducerPath]: moviesApi.reducer,
  [userApi.reducerPath]: userApi.reducer,
};

// Test helper
export const createTestStore = (preloadedState = {}) =>
  configureStore({
    reducer: reducers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(moviesApi.middleware)
        .concat(userApi.middleware),
    preloadedState,
  });

// --- DOM mocks for scrollable elements ---
export const setupDOMMocks = () => {
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "offsetWidth"
  );
  const originalScrollWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "scrollWidth"
  );

  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: 100,
    });
    Object.defineProperty(HTMLElement.prototype, "scrollWidth", {
      configurable: true,
      value: 500,
    });
  });

  afterAll(() => {
    if (originalOffsetWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "offsetWidth",
        originalOffsetWidth
      );
    if (originalScrollWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "scrollWidth",
        originalScrollWidth
      );
  });
};

// --- Unified render for components ---
interface RenderOptions {
  store?: ReturnType<typeof createTestStore>;
  wrapper?: React.ComponentType<{ children: ReactNode }>;
}

export const render = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (store) {
      return (
        <Provider store={store}>
          <MemoryRouter>{children}</MemoryRouter>
        </Provider>
      );
    }
    return <MemoryRouter>{children}</MemoryRouter>;
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

export const renderHookWithStore = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  {
    store,
    wrapper,
    ...options
  }: RenderHookOptions<TProps> & {
    store?: ReturnType<typeof createTestStore>;
  } = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (wrapper) return React.createElement(wrapper, null, children);
    return store ? (
      <Provider store={store}>
        <MemoryRouter>{children}</MemoryRouter>
      </Provider>
    ) : (
      <MemoryRouter>{children}</MemoryRouter> // always wrap in Router
    );
  };
  return rtlRenderHook(hook, { wrapper: Wrapper, ...options });
};

export const renderWithoutRouter = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return store ? (
      <Provider store={store}>{children}</Provider>
    ) : (
      <>{children}</>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};
