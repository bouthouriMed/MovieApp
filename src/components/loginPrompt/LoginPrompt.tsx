import React from "react";
import "./LoginPrompt.scss";
import { useTMDBAuth } from "../../hooks/useTMDBAuth";

const LoginPrompt: React.FC = () => {
  const { login } = useTMDBAuth();
  return (
    <div className="login-prompt">
      <div className="card">
        <h2>ðŸ‘€ Your Watchlist Awaits!</h2>
        <p>
          You can only see your watchlist if youâ€™re logged in with your TMDB
          account.
        </p>
        <button className="login-btn" onClick={login}>
          Login with TMDB
        </button>
      </div>
    </div>
  );
};

export default LoginPrompt;
