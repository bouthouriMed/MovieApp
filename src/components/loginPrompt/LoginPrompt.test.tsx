import { screen, fireEvent } from "@testing-library/react";
import { render } from "@/test-utils";
import LoginPrompt from "./LoginPrompt";

// Mock the useTMDBAuth hook
const mockLogin = vi.fn();

vi.mock("@/hooks/useTMDBAuth", () => ({
  useTMDBAuth: () => ({
    login: mockLogin,
  }),
}));

describe("LoginPrompt", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("matches snapshot", () => {
    const { container } = render(<LoginPrompt />);
    expect(container).toMatchSnapshot();
  });

  it("calls login when button is clicked", () => {
    render(<LoginPrompt />);
    const button = screen.getByRole("button", { name: /Login with TMDB/i });
    fireEvent.click(button);
    expect(mockLogin).toHaveBeenCalledTimes(1);
  });
});
