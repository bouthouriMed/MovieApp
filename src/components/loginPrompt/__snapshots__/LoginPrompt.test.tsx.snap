// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPrompt > matches snapshot 1`] = `
<div>
  <div
    class="login-prompt"
  >
    <div
      class="card"
    >
      <h2>
        ðŸ‘€ Your Watchlist Awaits!
      </h2>
      <p>
        You can only see your watchlist if youâ€™re logged in with your TMDB account.
      </p>
      <button
        class="login-btn"
      >
        Login with TMDB
      </button>
    </div>
  </div>
</div>
`;
